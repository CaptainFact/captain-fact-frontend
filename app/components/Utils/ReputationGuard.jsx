import React from "react"
import { connect } from "react-redux"
import { hasReputation } from '../../state/users/current_user/selectors'
import { LoadingFrame } from "."
import { ErrorView } from './ErrorView'


export const ReputationGuard = ({isLoading, hasReputation, showLoading, showNotEnough, children}) => {
  if (showLoading && isLoading)
    return <LoadingFrame/>
  if (hasReputation)
    return children
  return showNotEnough ? <ErrorView error="notEnoughReputation"/> : null
}

export default connect((state, props) => ({
  hasReputation: hasReputation(state, props.requiredRep),
  isLoading: state.CurrentUser.isLoading
}))(ReputationGuard)