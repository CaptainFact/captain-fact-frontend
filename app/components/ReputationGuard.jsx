import React from "react"
import { connect } from "react-redux"
import { translate } from 'react-i18next'
import { isAuthenticated, hasReputation } from '../state/users/current_user/selectors'
import { LoadingFrame, LinkWithIcon } from "./Utils"

@connect((state, props) => ({
  isAuthenticated: isAuthenticated(state),
  hasReputation: hasReputation(state, props.requiredRep),
  isLoading: state.CurrentUser.isLoading
}))
@translate('errors')
export default class ReputationGuard extends React.PureComponent {
  render() {
    const { isLoading, hasReputation, t } = this.props

    if (isLoading)
      return <LoadingFrame/>
    if (hasReputation)
      return this.props.children
    return <div>{ t('client.notEnoughReputation') }</div>
    }
}