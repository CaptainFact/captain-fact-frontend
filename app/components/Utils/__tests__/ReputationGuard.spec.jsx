import { ReputationGuard } from '../ReputationGuard'

test("should not render anything if no permissions until specified otherwise", () => {
  const noReputation = shallow(
    <ReputationGuard hasReputation={false}>Hidden</ReputationGuard>
  )
  const loading = shallow(
    <ReputationGuard isLoading={true}>Hidden</ReputationGuard>
  )
  expect(noReputation.isEmptyRender()).toBeTruthy()
  expect(loading.isEmptyRender()).toBeTruthy()
})

test("should not crash if no children", () => {
  expect(shallow(<ReputationGuard/>).isEmptyRender()).toBeTruthy()
})

test("should render children if allowed to", () => {
  snapshot(<ReputationGuard hasReputation>HelloWorld</ReputationGuard>)
})

test("display loading", () => {
  snapshot(<ReputationGuard isLoading showLoading>HelloWorld</ReputationGuard>)
})

test("show error", () => {
  snapshot(<ReputationGuard hasReputation={false} showNotEnough>HelloWorld</ReputationGuard>)
})