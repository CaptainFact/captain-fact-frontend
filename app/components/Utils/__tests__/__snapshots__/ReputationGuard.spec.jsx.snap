// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display loading 1`] = `<Translate(LoadingFrame) />`;

exports[`should render children if allowed to 1`] = `"HelloWorld"`;

exports[`show error 1`] = `
<withRouter(Translate(ErrorView))
  error="notEnoughReputation"
/>
`;
